modyn/selector/internal/selector_strategies/downsampling_strategies/rho_loss_downsampling_strategy.py

```diff
@@ -18,11 +25,53 @@ def inform_next_trigger(self, next_trigger_id: int, selector_storage_backend: Ab
         #
         # Will create a new table (pipeline_id, rho_pipeline_id) representing the mapping
         # between pipeline_id and rho_pipeline_id, primary key being pipeline_id.
-        #
+        rho_pipeline_id = self._get_or_create_rho_pipeline_id()
         # Step 2: Prepare the training data for the IL model, by randomly sampling a predefined ratio of samples
         # from next_trigger_id's data and storing them as a TSS with identifier (rho_pipeline_id, next_trigger_id).
-
+        self._prepare_holdout_set(next_trigger_id, rho_pipeline_id, selector_storage_backend)
         # Step 3: Issue training request to the trainer server, with pipeline_id as rho_pipeline_id and trigger_id
         # as next_trigger_id. Wait for the training to complete. Store the model. Record model id in
         # downsampling_params, so that it can be fetched and used for downsampling.
         raise NotImplementedError
+
+    def _get_or_create_rho_pipeline_id(self) -> int:
+        raise NotImplementedError
+
+    def _prepare_holdout_set(
+        self, next_trigger_id: int, rho_pipeline_id: int, selector_storage_backend: AbstractStorageBackend
+    ) -> None:
+        current_trigger_dataset_size = get_trigger_dataset_size(
+            selector_storage_backend, self._pipeline_id, next_trigger_id, tail_triggers=0
+        )
+
+        holdout_set_size = max(int(current_trigger_dataset_size * self.holdout_set_ratio / 100), 1)
+
+        stmt = self._get_holdout_sampling_query(self._pipeline_id, next_trigger_id, holdout_set_size).execution_options(
+            yield_per=self.maximum_keys_in_memory
+        )
+
+        def training_set_producer() -> Iterable[tuple[list[tuple[int, float]], dict[str, Any]]]:
```

Comment:
We need to assert with this implementation that the `DatabaseStorageBackend` is used since this implementation is not agnostic to the backend.

Reply:
Thanks for pointing it out! `_prepare_holdout_set` is an internal method called by the `inform_next_trigger`, where the assertion is already made.

Reply:
Ah, can you move the assertion to the constructor? I think that's where we typically assert this in the implementations

Reply:
The constructor has to access to this storage backend. The backend is directly passed to `inform_next_trigger`, see
https://github.com/eth-easl/modyn/blob/63be24037c82397a6160ba6d023732e37e683ee8/modyn/selector/internal/selector_strategies/downsampling_strategies/rho_loss_downsampling_strategy.py#L19

Reply:
Ah, I see. Thanks.